File: game/scripts/vscripts/abilities/heroes/queen/queen_wave/queen_wave.lua
queen_wave = class({})

function queen_wave:GetCastAnimationCustom() return ACT_DOTA_CAST_ABILITY_4 end
function queen_wave:GetPlaybackRateOverride() return 0.8 end
function queen_wave:GetCastPointSpeed() return 0 end

function queen_wave:OnAbilityPhaseStart()
	EFX('particles/units/heroes/hero_dawnbreaker/dawnbreaker_converge.vpcf', PATTACH_ABSORIGIN_FOLLOW, self:GetCaster(), { relese = true })
	return true
end

function queen_wave:OnSpellStart()
    local caster = self:GetCaster()
	local origin = caster:GetOrigin()
	local point = self:GetCursorPosition()
	local direction = Direction2D(origin, point)
	local projectile_speed = 850--self:GetSpecialValueFor("projectile_speed")
	local damage = self:GetSpecialValueFor("ability_damage")
	local damage_table = {
        damage = damage,
		damage_type = DAMAGE_TYPE_MAGICAL,
	}

	CustomEntities:ProjectileAttack(caster, {
		tProjectile = {
			EffectName = "particles/units/heroes/hero_queenofpain/queen_sonic_wave.vpcf",
			vSpawnOrigin = origin + Vector(direction.x * 45, direction.y * 45, 96),
			fDistance = 1000,--self:GetSpecialValueFor("projectile_distance"),
			fStartRadius = self:GetSpecialValueFor("hitbox"),
			fEndRadius = self:GetSpecialValueFor("hitbox") * 4,
			Source = caster,
			vVelocity = direction * projectile_speed,
			UnitBehavior = PROJECTILES_NOTHING,
			TreeBehavior = PROJECTILES_NOTHING,
			WallBehavior = PROJECTILES_DESTROY,
			GroundBehavior = PROJECTILES_NOTHING,
			fGroundOffset = 0,
			UnitTest = function(_self, unit) return unit:GetUnitName() ~= "npc_dummy_unit" and not CustomEntities:Allies(_self.Source, unit) end,
			OnUnitHit = function(_self, unit)
				damage_table.victim = unit
				damage_table.attacker = _self.Source
				
				ApplyDamage(damage_table)
				self:PlayEffectsOnHit(unit)
			end
		}
	})

	self:PlayEffectsOnCast()
end

function queen_wave:PlayEffectsOnCast()
	EmitSoundOn("Hero_QueenOfPain.SonicWave", self:GetCaster())
end

function queen_wave:PlayEffectsOnFinish(pos, particle_cast)
	local caster = self:GetCaster()
	EmitSoundOnLocationWithCaster(pos, "Hero_Centaur.DoubleEdge", caster)
	EmitSoundOnLocationWithCaster(pos, "Hero_Centaur.DoubleEdge.TI9_layer", caster)
	EmitSoundOnLocationWithCaster(pos, "Hero_Centaur.Gore", caster)

	local effect_cast = ParticleManager:CreateParticle(particle_cast, PATTACH_ABSORIGIN, caster)
	ParticleManager:SetParticleControl(effect_cast, 0, pos)
	ParticleManager:SetParticleControl(effect_cast, 1, pos)
	ParticleManager:SetParticleControl(effect_cast, 3, pos)
	
	ParticleManager:ReleaseParticleIndex(effect_cast)
end

function queen_wave:PlayEffectsOnHit(hTarget)
	local caster = self:GetCaster()
	EmitSoundOn("Hero_Sniper.AssassinateDamage", caster)

	local particle_cast = "particles/units/heroes/hero_sniper/sniper_assassinate_impact_blood.vpcf"
	local effect_cast = ParticleManager:CreateParticle(particle_cast, PATTACH_ABSORIGIN_FOLLOW, hTarget)

	ParticleManager:SetParticleControl(effect_cast, 0, hTarget:GetOrigin())
	ParticleManager:SetParticleControl(effect_cast, 1, hTarget:GetOrigin())
	ParticleManager:ReleaseParticleIndex(effect_cast)
end

if IsClient() then require("wrappers/abilities") end
Abilities.Castpoint(queen_wave)

File: src/vscripts/pve/abilities/heroes/queen/queen_wave.ts
import { CustomAbility } from "../../../../abilities/framework/custom_ability";
import { registerAbility } from "../../../../lib/dota_ts_adapter";
import { ProjectileBehavior } from "../../../../projectiles";
import { areUnitsAllied, direction2D } from "../../../../util";

@registerAbility("queen_wave")
class QueenWave extends CustomAbility {
    GetAnimation() {
        return GameActivity.DOTA_CAST_ABILITY_4;
    }

    GetPlaybackRateOverride() {
        return 0.8;
    }

    GetCastingCrawl() {
        return 0;
    }

    OnAbilityPhaseStart() {
        if (super.OnAbilityPhaseStart()) {
            EFX("particles/units/heroes/hero_dawnbreaker/dawnbreaker_converge.vpcf", ParticleAttachment.ABSORIGIN_FOLLOW, this.caster, {
                relese: true
            });
            return true;
        }
        return false;
    }

    OnSpellStart() {
        const origin = this.caster.GetOrigin();
        const point = this.GetCursorPosition();
        const projectileDirection = direction2D(origin, point);
        const projectileSpeed = 850;
        const projectileDistance = 1000;
        const damage = this.GetSpecialValueFor("ability_damage");

        this.ProjectileAttack({
            source: this.caster,
            effectName: "particles/units/heroes/hero_queenofpain/queen_sonic_wave.vpcf",
            distance: projectileDistance,
            startRadius: this.GetSpecialValueFor("hitbox"),
            endRadius: this.GetSpecialValueFor("hitbox") * 4,
            spawnOrigin: origin.__add(Vector(projectileDirection.x * 45, projectileDirection.y * 45, 96)),
            velocity: projectileDirection.__mul(projectileSpeed),
            unitBehavior: ProjectileBehavior.NOTHING,
            groundOffset: 0,
            unitTest: (unit, projectile) => !areUnitsAllied(projectile.getSource(), unit),
            onUnitHit: (unit, projectile) => {
                ApplyDamage({
                    victim: unit,
                    attacker: projectile.getSource(),
                    damage: damage,
                    damage_type: DamageTypes.MAGICAL
                });
                this.PlayEffectsOnHit(unit);
            }
        });

        EmitSoundOn("Hero_QueenOfPain.SonicWave", this.caster);
    }

    PlayEffectsOnFinish(position: Vector, path: string) {
        EmitSoundOnLocationWithCaster(position, "Hero_Centaur.DoubleEdge", this.caster);
        EmitSoundOnLocationWithCaster(position, "Hero_Centaur.DoubleEdge.TI9_layer", this.caster);
        EmitSoundOnLocationWithCaster(position, "Hero_Centaur.Gore", this.caster);

        const particleId = ParticleManager.CreateParticle(path, ParticleAttachment.ABSORIGIN, this.caster);
        ParticleManager.SetParticleControl(particleId, 0, position);
        ParticleManager.SetParticleControl(particleId, 1, position);
        ParticleManager.SetParticleControl(particleId, 3, position);
        ParticleManager.ReleaseParticleIndex(particleId);
    }

    PlayEffectsOnHit(target: CDOTA_BaseNPC) {
        EmitSoundOn("Hero_Sniper.AssassinateDamage", this.caster);

        const particleId = ParticleManager.CreateParticle(
            "particles/units/heroes/hero_sniper/sniper_assassinate_impact_blood.vpcf",
            ParticleAttachment.ABSORIGIN_FOLLOW,
            target
        );

        ParticleManager.SetParticleControl(particleId, 0, target.GetAbsOrigin());
        ParticleManager.SetParticleControl(particleId, 1, target.GetAbsOrigin());
        ParticleManager.ReleaseParticleIndex(particleId);
    }
}
